10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,107 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{store: storer}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+	} else {
+		puppy, readErr := rh.store.ReadPuppy(uint32(id))
+		if readErr != nil {
+			w.WriteHeader(http.StatusNotFound)
+			http.Error(w, http.StatusText(404)+": "+ErrorReadPuppyValue.String(), 404)
+		} else {
+			render.JSON(w, r, puppy)
+		}
+		return
+	}
+}
+
+// HandlePost post a new puppy and then display the results.
+func (rh *RestHandler) HandlePost(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if decodeErr := render.DecodeJSON(r.Body, &p); decodeErr != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidJSONInput.String(), 400)
+	} else {
+		_, createErr := rh.store.CreatePuppy(&p)
+		if createErr != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			render.JSON(w, r, createErr)
+		} else {
+			render.JSON(w, r, p)
+		}
+		return
+	}
+}
+
+// HandlePut updates the existing puppy with new fields in the payload.
+func (rh *RestHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
```

Comment:
Hey,

This is not working as expected. Please consider following requests.

```
 niemiram@niemiram-JSS1271  /private/tmp/go-course/10_rest/willshen8   lab10 ●  curl http://127.0.0.1:8888/api/puppy/4 -X PUT -d '{"value":"-42"}' -i
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Sun, 25 Aug 2019 23:35:00 GMT
Content-Length: 5
Proxy-Connection: keep-alive
Connection: keep-alive

true
(⎈ :)
 niemiram@niemiram-JSS1271  /private/tmp/go-course/10_rest/willshen8   lab10 ●  curl http://127.0.0.1:8888/api/puppy/4
{"id":4,"breed":"","color":"","value":"-42"}
(⎈ :)
```
and
```
 niemiram@niemiram-JSS1271  /private/tmp/go-course/10_rest/willshen8   lab10 ●  curl http://127.0.0.1:8888/api/puppy/444 -X PUT -d '{"value":"-42"}' -i
HTTP/1.1 422 Unprocessable Entity
Date: Sun, 25 Aug 2019 23:35:15 GMT
Content-Length: 74
Content-Type: text/plain; charset=utf-8
Proxy-Connection: keep-alive
Connection: keep-alive

Unprocessable Entity: Puppy ID can not be found, update operation failed.
```

Reply:
I need to have this reviewed by Julia today, so please don't change the status. I will look into it. Bear in mind that this is an academic exercise, so I won't get too hang up on thee statuses.

Reply:
I'm not changing any labels, so Julia or other Core Reviewer should review the code :)

Anyway I believe providing working http server is part of this exercise. I'd disagree that statuses are not important, but It's up to to core team to decide.

Reply:
I don't mean not important, I even changed POST and PUT to 'render.Status(r, http.StatusCreated)` for this reason. Doesn't normally the UI layer would handle exceptions such as entering negative numbers, and does all the checks to ensure user input is valid?

Reply:
please fix.

Reply:
@willshen8 
> "Doesn't normally the UI layer would handle exceptions such as entering negative numbers, and does all the checks to ensure user input is valid?" or providing valid requests.

This approach is 1st step to be hacked. User can send any request to the server. We cannot relay that user will be always using provided `JS` based website.